### PR TITLE
feat: add error extraction logic

### PR DESCRIPTION
**Pull request by sanpj2292**
# Description

~~Add error extraction logic using `gjson`~~
Updated the logic a bit used json in some parts for more concrete evaluation of json & parsing the json itself before we go for recursive search of probable error message keys

## Notion Ticket

https://www.notion.so/rudderstacks/Adding-errorMessage-field-in-reports-table-of-rudder-server-e584e48c6959476abc5b5b72dd8693c3?pvs=4

## Security

- [ ] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

**Comment by sanpj2292** [comment]
Since this is not the way we want to proceed, closing the PR

File: enterprise/reporting/error-detail-extractor.go [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+	if errors, ok := getValue([]string{errorsKey}, findKeys([]string{errorsKey}, resp)).([]interface{}); ok && len(errors) > 0 {
+		if message := findFirstExistingKey(ext.ErrorMessageKeys, resp); message != nil {
+			if s, ok := message.(string); ok {
+				return s
+			}
+		}
+	}
+	// split the error message by newline and take the first entry
+
+	// for _, whKey := range warehouseKeys {
```
Comment (koladilip) [inline enterprise/reporting/error-detail-extractor.go:131]: warehouse errors are considered?

File: enterprise/reporting/error-detail-extractor.go [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+	defaultErrorMessageKeys = append(defaultErrorMessageKeys, errMsgKeys...)
+	extractor := &ExtractorT{
+		ErrorMessageKeys: defaultErrorMessageKeys,
+		log:              log.Child("ErrorDetailExtractor"),
+	}
+	return extractor
+}
+
+// Functions used for error message extraction -- STARTS
+func isMapOrArray(value interface{}) bool {
```
[the remark this answers is not in the document]
Comment (koladilip) [reply enterprise/reporting/error-detail-extractor.go:43]: https://go.dev/play/p/qfSSXmJbXmp check this, this function may not work properly.

[the remark this answers is not in the document]
Comment (sanpj2292) [reply enterprise/reporting/error-detail-extractor.go]: It is working for us now. Can you once check and confirm ?

File: enterprise/reporting/error-detail-extractor.go (hunk 1 above)
Comment (sanpj2292) [reply enterprise/reporting/error-detail-extractor.go]: They are considered in the updated logic

File: enterprise/reporting/error-detail-extractor.go [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+	arrOfErrs, isIntfArr := errorsI.([]interface{})
+	if !isIntfArr {
+		return ""
+	}
+	errors := lo.Uniq(arrOfErrs)
+	errorStrings := make([]string, len(errors))
+	for i, v := range errors {
+		errorStrings[i] = fmt.Sprint(v)
+	}
+	return strings.Join(errorStrings, ".")
```
Comment (koladilip) [inline enterprise/reporting/error-detail-extractor.go:225]: This code is duplicated in skipToSome block

**Review by koladilip — APPROVED** [review]

**Comment by sanpj2292** [comment]
Merging to an internal branch.

**Pull request merged**